pkg/workloadmeta/collectors/internal/containerd/image.go

```diff
@@ -0,0 +1,344 @@
+		}
+
+		return c.handleImageCreateOrUpdate(ctx, containerdEvent.Namespace, event.Name)
+
+	case imageDeletionTopic:
+		event := &events.ImageDelete{}
+		if err := proto.Unmarshal(containerdEvent.Event.Value, event); err != nil {
+			return fmt.Errorf("error unmarshaling containerd event: %w", err)
+		}
+
+		imageID, found := c.knownImages[event.Name]
+		if !found {
+			// Not necessarily an error. If an image had multiple names, it
+			// could have already been deleted using other name
+			return nil
+		}
+
+		c.store.Notify([]workloadmeta.CollectorEvent{
+			{
+				Type:   workloadmeta.EventTypeUnset,
+				Source: workloadmeta.SourceRuntime,
+				Entity: &workloadmeta.ContainerImageMetadata{
+					EntityID: workloadmeta.EntityID{
+						Kind: workloadmeta.KindContainerImageMetadata,
+						ID:   imageID,
+					},
+				},
+			},
+		})
+
+		delete(c.knownImages, event.Name)
+
+		return nil
+	default:
+		return fmt.Errorf("unknown containerd image event topic %s, ignoring", containerdEvent.Topic)
+	}
+}
+
+func (c *collector) handleImageCreateOrUpdate(ctx context.Context, namespace string, imageName string) error {
+	img, err := c.containerdClient.Image(namespace, imageName)
+	if err != nil {
+		return fmt.Errorf("error getting image: %w", err)
+	}
+
+	return c.notifyEventForImage(ctx, namespace, img)
+}
+
+func (c *collector) notifyEventForImage(ctx context.Context, namespace string, img containerd.Image) error {
+	ctxWithNamespace := namespaces.WithNamespace(ctx, namespace)
+
+	manifest, err := images.Manifest(ctxWithNamespace, img.ContentStore(), img.Target(), img.Platform())
+	if err != nil {
+		return fmt.Errorf("error getting image manifest: %w", err)
+	}
+
+	layers, err := getLayersWithHistory(ctxWithNamespace, img.ContentStore(), manifest)
+	if err != nil {
+		_ = log.Warnf("error while getting layers with history: %s", err)
+
+		// Not sure if the layers and history are always available. Instead of
+		// returning an error, collect the image without this information.
+	}
+
+	platforms, err := images.Platforms(ctxWithNamespace, img.ContentStore(), manifest.Config)
+	if err != nil {
+		return fmt.Errorf("error getting image platforms: %w", err)
+	}
+
+	imageName := img.Name()
+	imageID := manifest.Config.Digest.String()
+
+	// We can get "create" events for images that already exist. That happens
+	// when the same image is referenced with different names. For example,
+	// datadog/agent:latest and datadog/agent:7 might refer to the same image.
+	// Also, in some environments (at least with Kind), pulling an image like
+	// datadog/agent:latest creates several events: in one of them the image
+	// name is a digest, in other is something with the same format as
+	// datadog/agent:7, and sometimes there's a temporary name prefixed with
+	// "import-".
+	existingImg, err := c.store.GetImage(imageID)
+	if err == nil {
+		updatedImg := existingImg.DeepCopy().(*workloadmeta.ContainerImageMetadata) // Avoid race conditions
```

Comment:
Is `existingImg` a reference to an object that can be modified by a concurrent goroutine ?
The comment `// Avoid race conditions` suggests that we need to protect ourselves against a concurrent modification.
But I can’t see any mutex lock inside the implementation of `DeepCopy()`.

https://github.com/DataDog/datadog-agent/blob/636b0f9603a0e892989c98fe4418a1adc7af6728/pkg/workloadmeta/types.go#L659-L662

https://github.com/mohae/deepcopy/blob/c48cc78d482608239f6c4c92a4abd87eb8761c90/deepcopy.go#L27-L43